File: src/main/java/kitchenpos/eatinorders/domain/tobe/domain/EatInOrderCompleteService.java
package kitchenpos.eatinorders.domain.tobe.domain;

import kitchenpos.eatinorders.domain.OrderStatus;
import kitchenpos.eatinorders.domain.tobe.domain.vo.Guests;
import kitchenpos.eatinorders.dto.EatInOrderResponse;
import kitchenpos.eatinorders.dto.OrderStatusChangeRequest;
import org.springframework.stereotype.Service;

import java.util.NoSuchElementException;

@Service
public class EatInOrderCompleteService {
    private final EatInOrderRepository orderRepository;
    private final TobeOrderTableRepository orderTableRepository;


    public EatInOrderCompleteService(EatInOrderRepository orderRepository, TobeOrderTableRepository orderTableRepository) {
        this.orderRepository = orderRepository;
        this.orderTableRepository = orderTableRepository;
    }

    public EatInOrderResponse complete(OrderStatusChangeRequest request) {
        EatInOrder order = orderRepository.findById(request.getOrderId())
                .orElseThrow(NoSuchElementException::new)
                .complete();
        TobeOrderTable table = orderTableRepository.findById(order.getOrderTableId())
                .orElseThrow(NoSuchElementException::new);
        if (!orderRepository.existsByOrderTableAndStatusNot(table, OrderStatus.COMPLETED)) {
            table.changeNumberOfGuests(new Guests(0));
            table.clear();
        }
        return new EatInOrderResponse(order);
    }
}

File: src/main/kotlin/kitchenpos/tobe/eatinorder/domain/EatInOrderCompleteService.kt
package kitchenpos.tobe.eatinorder.domain

import kitchenpos.order.common.domain.OrderStatus
import kitchenpos.tobe.eatinorder.domain.repository.EatInOrderRepository
import kitchenpos.tobe.support.DomainService
import java.util.*

@DomainService
class EatInOrderCompleteService(
    private val eatInOrderRepository: EatInOrderRepository,
) {
    /***
     * 주문 테이블의 모든 주문이 완료됐으면
     * 테이블을 초기화한다.
     */
    fun complete(orderId: UUID): UUID {
        val orderTable = eatInOrderRepository.findEatInOrderById(orderId).orderTable
        val isAllCompleted =
            eatInOrderRepository.existsByOrderTableAndStatusNot(
                orderTable,
                OrderStatus.COMPLETED,
            ).not()

        check(isAllCompleted) { "주문 테이블의 모든 주문이 완료되지 않았습니다." }

        val eatInOrder = eatInOrderRepository.findEatInOrderById(orderId)
        eatInOrder.complete()
        orderTable.clear()

        return eatInOrder.id
    }
}
